loader/loader.c

```diff
@@ -6852,356 +6912,334 @@ VkResult setup_loader_term_phys_dev_groups(struct loader_instance *inst) {
+                }
 
-            // Add each GPU as it's own group
-            for (uint32_t indiv_gpu = 0; indiv_gpu < count_this_time; indiv_gpu++) {
-                local_phys_dev_groups[indiv_gpu + cur_icd_group_count].this_icd_term = icd_term;
-                local_phys_dev_groups[indiv_gpu + cur_icd_group_count].icd_index = icd_idx;
-                local_phys_dev_groups[indiv_gpu + cur_icd_group_count].group_props.physicalDeviceCount = 1;
-                local_phys_dev_groups[indiv_gpu + cur_icd_group_count].group_props.physicalDevices[0] = phys_dev_array[indiv_gpu];
-                local_phys_dev_group_sorted[indiv_gpu + cur_icd_group_count] = icd_sorted;
-            }
+                // Add each GPU as it's own group
+                for (uint32_t indiv_gpu = 0; indiv_gpu < count_this_time; indiv_gpu++) {
+                    uint32_t cur_index = indiv_gpu + cur_icd_group_count;
+                    local_phys_dev_groups[cur_index].this_icd_term = icd_term;
+                    local_phys_dev_groups[cur_index].icd_index = icd_idx;
+                    local_phys_dev_groups[cur_index].group_props.physicalDeviceCount = 1;
+                    local_phys_dev_groups[cur_index].group_props.physicalDevices[0] = phys_dev_array[indiv_gpu];
+                }
 
-        } else {
-            res = fpEnumeratePhysicalDeviceGroups(icd_term->instance, &count_this_time,
-                                                  &local_phys_dev_groups[cur_icd_group_count].group_props);
-            for (uint32_t group = 0; group < count_this_time; ++group) {
-                local_phys_dev_group_sorted[group + cur_icd_group_count] = icd_sorted;
-                local_phys_dev_groups[group + cur_icd_group_count].this_icd_term = icd_term;
-                local_phys_dev_groups[group + cur_icd_group_count].icd_index = icd_idx;
-            }
-            if (VK_SUCCESS != res) {
-                loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
-                           "setup_loader_term_phys_dev_groups:  Failed during dispatch call of "
-                           "\'EnumeratePhysicalDeviceGroups\' to ICD %d to get content.",
-                           icd_idx);
-                goto out;
+            } else {
+                res = fpEnumeratePhysicalDeviceGroups(icd_term->instance, &count_this_time, NULL);
+                if (res != VK_SUCCESS) {
+                    loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
+                               "terminator_EnumeratePhysicalDeviceGroups:  Failed during dispatch call of "
+                               "\'EnumeratePhysicalDeviceGroups\' to ICD %d to get group count.",
+                               icd_idx);
+                    goto out;
+                }
+                if (cur_icd_group_count + count_this_time < *pPhysicalDeviceGroupCount) {
+                    // Can just use passed in structs
+                    res = fpEnumeratePhysicalDeviceGroups(icd_term->instance, &count_this_time,
```

Comment:
Good to better describe how we can use the passed in pointer for storage. Was confused at first where `pPhysicalDeviceGroupProperties` came from.

Reply:
Added more comments